nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java

```diff
@@ -0,0 +1,352 @@
+                    Path path = workDir.resolve(fileName);
+            
+                    try {
+                        // new directory to be watched and traversed
+                        if (kind == ENTRY_CREATE && Files.isDirectory(path) && !Files.isHidden(path) && Files.isReadable(path)) {
+                            registerAll(path);
+                            initializePropertiesFromPath(path);
+                        }
+                        // new or updated file found
+                        if (Files.isRegularFile(path) && (kind == ENTRY_CREATE || kind == ENTRY_MODIFY)) {
+                            updatePropertyFromPath(path, Files.readAttributes(path, BasicFileAttributes.class));
+                        }
+                        if (Files.isRegularFile(path) && (kind == ENTRY_DELETE)) {
+                            properties.remove(parsePropertyNameFromPath(path));
+                        }
+                    } catch (IOException e) {
+                        logger.log(Level.SEVERE, "Could not process event '"+kind+"' on '"+path+"'", e);
+                    }
+                }
+    
+                // Reset key (obligatory) and remove from set if directory no longer accessible
+                boolean valid = key.reset();
+                if (!valid) {
+                    keys.remove(key);
+                    
+                    // all directories became inaccessible, even topmostDir!
+                    if (keys.isEmpty()) break;
+                }
+            }
+        }
+    }
+    
+    class DirConfigFileVisitor extends SimpleFileVisitor<Path> {
+        /**
+         * Ignore hidden directories
+         */
+        @Override
+        public FileVisitResult preVisitDirectory(java.nio.file.Path dir, BasicFileAttributes attrs) throws IOException {
+            return dir.toFile().isHidden() ? FileVisitResult.SKIP_SUBTREE : FileVisitResult.CONTINUE;
+        }
+    
+        @Override
+        public FileVisitResult visitFile(Path path, BasicFileAttributes mainAtts) throws IOException {
+            updatePropertyFromPath(path, mainAtts);
+            return FileVisitResult.CONTINUE;
+        }
+    }
+    
+    private Path directory;
+    private ConcurrentHashMap<String, DirProperty> properties = new ConcurrentHashMap<>();
+    private static Logger logger = Logger.getLogger(DirConfigSource.class.getName());
+    
+    public DirConfigSource(PayaraExecutorService executorService) {
+        // get the directory from the app server config
+        findDir();
+        
+        try {
+            // create the watcher for the directory
+            executorService.submit(new DirPropertyWatcher(this.directory));
+            // initial loading
+            initializePropertiesFromPath(this.directory);
+        } catch (IOException e) {
+            logger.log(Level.SEVERE, "Error during setup of MicroProfile Config Directory Source", e);
+        }
+    }
```

Comment:
This might as well be moved to an anonymous definition inside `initializePropertiesFromPath()`. Had been moved here for potential reuse, but didn't happen.